cis194/week2/robert-chang/src/LogAnalysis.hs

```diff
@@ -53,3 +53,48 @@ parse :: String -> [LogMessage]
 parse [] = []
 parse file = map parseMessage $ lines file
 
+
+
+-- Exercise 2
+
+insert :: LogMessage -> MessageTree -> MessageTree
+{- if insert is given a LogMessage which is
+Unknown, it should return the MessageTree unchanged
+-}
+{- insert to empty tree, should have LogMessage inserted -} 
+insert x Leaf        = Node Leaf x Leaf 
+insert x@(LogMessage _ t1 _) (Node left xs@(LogMessage _ t2 _) right) 
+  | t1 == t2 = Node left xs right
+  | t1 < t2  = Node (insert x left) xs right
+  | t1 > t2  = Node left xs (insert x right)
+
+insert _ tree = tree
+
+-- Exercise 3
+build :: [LogMessage] -> MessageTree
+-- foldr :: (a -> b -> b) -> b -> t a -> b
+-- e.g.
+-- foldr f z []     = z 
+-- foldr f z (x:xs) = f x (foldr f z xs)
+-- In this case:
+-- foldr insert Leaf [x1, x2, ..., xn] == x1 `insert` (x2 `insert` ... (xn `insert` Leaf)...)
+build = foldr insert Leaf 
+
+-- Exercise 4
+inOrder :: MessageTree -> [LogMessage]
+inOrder Leaf = []
+inOrder (Node left x right) = inOrder left ++ [x] ++ inOrder right
+
+-- Exercise 5
+whatWentWrong :: [LogMessage] -> [String]
+whatWentWrong =  fmap getMsg . inOrder . build . filter . severity $ 50
+
+severity :: Int -> LogMessage -> Bool
+severity level (LogMessage (Error lvl) _ _)
+  | lvl >= level = True
```

Comment:
You don't need to write `lvl >= level = True` since `lvl >= level` returns a `Bool`